cpp/include/cudf/groupby.hpp

```diff
@@ -19,47 +19,175 @@
+ * specified by a list of operator enums of equal length to the number of value
+ * columns.
+ *
+ * The output of the operation is the table of key columns that hold all the
+ * unique keys from the input key columns and a table of aggregation columns
+ * that hold the specified reduction across associated values among all
+ * identical keys.
+ *
+ * @param keys The table of keys
+ * @param values The table of aggregation values
+ * @param ops The list of aggregation operations
+ * @return A tuple whose first member contains the table of output keys, and
+ * second member contains the table of reduced output values
+ *---------------------------------------------------------------------------**/
+std::pair<cudf::table, cudf::table> groupby(cudf::table const& keys,
+                                            cudf::table const& values,
+                                            std::vector<operators> const& ops,
+                                            Options options = Options{});
+
+}  // namespace hash
+
+namespace sort {
+
+/**---------------------------------------------------------------------------*
+ * @brief  Options unique to the sort-based groupby
+ *
+ *---------------------------------------------------------------------------**/
+struct Options : groupby::Options {};
+
+/**---------------------------------------------------------------------------*
+ * @brief Supported aggregation operations
+ *
+ *---------------------------------------------------------------------------**/
+enum operators { SUM, MIN, MAX, COUNT };
+
+/**---------------------------------------------------------------------------*
+ * @brief Performs groupby operation(s) via a sort-based implementation
+ *
+ * Given a table of keys and corresponding table of values, equivalent keys will
+ * be grouped together and a reduction operation performed across the associated
+ * values (i.e., reduce by key). The reduction operation to be performed is
+ * specified by a list of operator enums of equal length to the number of value
+ * columns.
+ *
+ * The output of the operation is the table of key columns that hold all the
+ * unique keys from the input key columns and a table of aggregation columns
+ * that hold the specified reduction across associated values among all
+ * identical keys.
+ *
+ * @param keys The table of keys
+ * @param values The table of aggregation values
+ * @param ops The list of aggregation operations
+ * @return A tuple whose first member contains the table of output keys, and
+ * second member contains the table of reduced output values
+ *---------------------------------------------------------------------------**/
+std::pair<cudf::table, cudf::table> groupby(cudf::table const& keys,
+                                            cudf::table const& values,
+                                            std::vector<operators> const& ops,
+                                            Options options);
```

Comment:
If you explicitly refer to it as `groupby::options` here can you use the same spelling for the struct name and the parameter name?

```suggestion
                                            groupby::options options);
```

Reply:
You'd have to do `groupby::hash::options`, and you'd have to refer to it that way in _all_ function declarations. Way too verbose for my liking. 

Alternatives:
1. Keep `Options`
2. `options_t`
3. Use `options` and name parameter `opts`

2. isn't great because `*_t` is reserved and is verboten.
3. Is worse because `opts` and `ops` are far too close together.

Reply:
I tried `hash::groupby::options options`, and it doesn't work. The compiler seems to get confused:

```
../include/cudf/groupby.hpp:112:140: error: expected primary-expression before ‘struct’
../include/cudf/groupby.hpp:112:140: error: expected ‘,’ or ‘...’ before ‘struct’
```